### PR TITLE
[stdlib] Initial FreeBSD port.

### PR DESCRIPTION
**Pull request by dcci**
Initial FreeBSD port of stdlib/FreeBSDlibc.
Thanks!
## 

Davide

### CODE REVIEW

File: stdlib/private/CMakeLists.txt [hunk 1]
```diff
@@ -19,3 +19,9 @@ if(CMAKE_SYSTEM_NAME STREQUAL "Linux")
   add_subdirectory(SwiftPrivateDarwinExtras)
   add_subdirectory(SwiftPrivatePthreadExtras)
 endif()
+
+if(CMAKE_SYSTEM_NAME STREQUAL "FreeBSD")
```
Comment (gribozavr) [inline stdlib/private/CMakeLists.txt]: Any reason to not do an 'or' and add this to the previous 'if' statement?  I expect these three libraries to be buildable on any unix-y system.

File: stdlib/public/FreeBSDlibc/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,41 @@
+set(sources
+    FreeBSDlibc.swift
```
Comment (gribozavr) [inline stdlib/public/FreeBSDlibc/CMakeLists.txt]: Do you really want to call it `FreeBSDlibc`?  Why not just `libc`?

File: stdlib/public/FreeBSDlibc/FreeBSDlibc.swift [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+  return _swift_FreeBSDlibc_sem_open4(name, oflag, mode, value)
+}
+
+// extern char **environ;
+@_silgen_name("_swift_FreeBSDlibc_getEnvironment")
+internal func _swift_FreeBSDlibc_getEnvironment(
+) -> UnsafeMutablePointer<UnsafeMutablePointer<CChar>>
+
+@warn_unused_result
+public func freebsdGetEnviron() -> UnsafeMutablePointer<UnsafeMutablePointer<CChar>> {
```
Comment (gribozavr) [inline stdlib/public/FreeBSDlibc/FreeBSDlibc.swift]: If `extern char **environ` is the API on FreeBSD that we don't import, then the overlay should be exposing a readonly var:

```
var environ: UnsafeMutablePointer<UnsafeMutablePointer<CChar>> {
  ...
}
```

**Comment by gribozavr** [comment]
@dcci Thank you for your patch!

I don't like the duplication of the overlay code.  Is there a way you can think of we could share the code between FreeBSD and Linux, and, potentially, Darwin?

File: stdlib/private/StdlibUnittest/StdlibUnittest.swift.gyb [hunk 4]
```diff
@@ -1005,6 +1008,8 @@ public enum OSVersion : CustomStringConvertible {
       return "watchOSSimulator"
     case Linux:
       return "Linux"
+    case FreeBSD:
+      return "FreeBSD"
```
Comment (gribozavr) [inline stdlib/private/StdlibUnittest/StdlibUnittest.swift.gyb]: There is also `TestRunPredicate` below that should be updated, and then you should be able to write a test like this `validation-test/stdlib/StdlibUnittestLinux.swift` but for FreeBSD.

File: stdlib/public/FreeBSDlibc/FreeBSDlibc.swift [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+#endif
+
+//===----------------------------------------------------------------------===//
+// semaphore.h
+//===----------------------------------------------------------------------===//
+
+/// The value returned by `sem_open()` in the case of failure.
+public var SEM_FAILED: UnsafeMutablePointer<sem_t> {
+  // The value is ABI.  Value verified to be correct on FreeBSDlibc.
+  return UnsafeMutablePointer<sem_t>(bitPattern: 0)
```
Comment (gribozavr) [inline stdlib/public/FreeBSDlibc/FreeBSDlibc.swift]: Sorry, just to double-check, did you really verify this value?

File: stdlib/public/FreeBSDlibc/module.map.in [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+/// This is a semi-complete modulemap that maps glibc's headers in a roughly
+/// similar way to the Darwin SDK modulemap. We do not take care to list every
+/// single header which may be included by a particular submodule, so there can
+/// still be issues if imported into the same context as one in which someone
+/// included those headers directly.
+///
+/// It's not named just Glibc so that it doesn't conflict in the event of a
```
Comment (gribozavr) [inline stdlib/public/FreeBSDlibc/module.map.in]: Glibc?

Comment (dcci) [inline stdlib/public/FreeBSDlibc/module.map.in]: Oversight, I'll fix.

File: stdlib/public/FreeBSDlibc/FreeBSDlibc.swift (hunk 5 above)
Comment (dcci) [inline stdlib/public/FreeBSDlibc/FreeBSDlibc.swift]: Yes they seem to match. I'll check anyway.

File: stdlib/private/StdlibUnittest/StdlibUnittest.swift.gyb (hunk 4 above)
Comment (dcci) [inline stdlib/private/StdlibUnittest/StdlibUnittest.swift.gyb]: I see, thanks!

**Comment by dcci** [comment]
The main difference is the module map and __environ. There's probably a way to make more code to be shared, I'll think about it harder between today and tomorrow and submit a new patch

Thanks for your review!

--
Davide

**Comment by gribozavr** [comment]
@dcci Thanks!  I would be much happier with more complex CMake code than duplication of actual overlay code.

**Pull request closed without merging**